# Dynamic Programming

## Purpose

The simplest way to think about dynamic programming, "DP", is nothing more than as an approach to optimize an algorithm:

1.  Find an algorithm that "brute forces" a problem
2.  Notice that it fits into the the class of algorithms DP can optimize
3.  Implement algorithm using DP

Dynamic programming does **not** have anything to do with on-the-fly code generation, online reinforcement learning, editing/optimizing compiled code while it's running... or anything else fancy.  So what does it do?  Read on!

## Approach

It boils down to a simple idea:  _don't let your code do the same work twice_.  In practice, this means that you should cache intermediate results (say, in a table).  Each time you need to evaluate a function with the same input after the first, you can look up the result in the cache rather than recomputing it.

You will sometimes hear DP is used for problems whose solutions are combinations of the solutions to "optimal subproblems".  Unpacking this a little bit:  if each solution is just a combination of smaller versions of the same problem, and you cache the results of the smaller problems as you go, to the extent that you need those solutions in more than one place, you'll get a speed up.  For many problems, an exponential time naive implementation can be made to run in polynomial time.

**Don't let this sound complicated to you.**  All we're doing is adding a cache to avoid recomputing an "expensive" function.  Any time you see a problem that sounds like, "Try every possible combination of X and pick the best", you should think "I wonder if there are any partial results here we can cache and reuse".

## Why is this important for natural language processing?

In the second half of the course, we'll study tasks where the desired output is a sequence, a tree, or another type of rich structure. Unlike simple classification problems, we can't easily enumerate all the possible outputs - because there are exponentially many!

In many such tasks, we can still decompose them into sub-problems, like predicting a single branch of a tree. Then by using dynamic programming, we can devise an efficient algorithm to find the highest-scoring solution. We'll make heavy use of this approach when we study parsing (week 9-10) and sequence labeling tasks like part-of-speech tagging (week 7-8).

Dynamic programming is also useful as a conceptual tool. Decomposing a large problem into subproblems can be as useful for us as for computers, since it lets us reason about small, controlled cases and extend them into a more complex solution. Dynamic programming is also closely related to recursion - a concept underlying many fundamental ideas in both computer science and linguistics.

## Example: Climbing the stairs

Let's say you want to compute the total number of ways you can run up a flight of stairs.  At each step $k$ you can either:

- Take a regular step up to $k + 1$
- Take a big step up to $k + 2$.

Starting on stair $0$ (the bottom floor), how many unique ways are there to get up a staircase with $n$ stairs?

If you've not encountered this kind of problem before, **spend a minute pondering** how you might compute this.  It's common to feel "lost" in the combinatorial explosion of options.

### Hint

Could you figure out the number of ways up a staircase of height $n$ if you were told how many ways you might get up staircases of height $n - 1$ and $n - 2$?  That is, if you knew the solution to a _subproblem_, could you use that to compute the solution to your real problem?

### Further Hint

Say you can get to the second-to-the-top stair in 8 ways (and then take a big step) and the next-to-the-top stair in 13 ways (and then take a small step).  How many ways to the top?  8 + 13 = 21.  (Note that you don't have to count "get to second-to-the-top and take two regular steps" separately because the "13" already includes them.)

### Solution

In general, if there are $ways(n - 1)$ to get to the $n - 1$ stair, you can use any of those methods to get there and then take a regular step to get to stair $n$.  Similarly, if you have $ways(n - 2)$ to get to the $n - 2$ step and then take a big step to stair $n$.  There is no other way to get to stair $n$ except by one of those two sets of options.  There is no overlap between those sets of sequences as sequences in the first set always end in a "regular" step and those sequences in the second set always end in a "big" step.

In math, $ways(n) = ways(n-1) + ways(n-2)$.

Also notice (imagine, or draw a picture), $ways(1) = 1$ and $ways(2) = 2$.

*Aside: you may recognize this as the famous [Fibonacci series](https://en.wikipedia.org/wiki/Fibonacci_number).*

The next cell shows one implementation.

In [1]:
def naive_ways(n):
    """Return the ways up a staircase of length n. Uses a naive algorithm."""
    if n == 1:
        return 1
    if n == 2:
        return 2
    
    return naive_ways(n - 1) + naive_ways(n - 2)

naive_ways(4)

5

**Great!**

Unfortunately, this implementation gets very slow for large $n$. The cell below will print some timing information.

In [17]:
# TIMING INFO FOR PART A
time_list = []
for n in range(20, 25+1):
    print("n=%d: " % n),
    %timeit -n25 naive_ways(n)

n=20: 
2.8 ms ± 810 µs per loop (mean ± std. dev. of 7 runs, 25 loops each)
n=21: 
3.89 ms ± 530 µs per loop (mean ± std. dev. of 7 runs, 25 loops each)
n=22: 
5.77 ms ± 167 µs per loop (mean ± std. dev. of 7 runs, 25 loops each)
n=23: 
9.75 ms ± 1.26 ms per loop (mean ± std. dev. of 7 runs, 25 loops each)
n=24: 
16.2 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 25 loops each)
n=25: 
28.7 ms ± 3.38 ms per loop (mean ± std. dev. of 7 runs, 25 loops each)


In [3]:
# Try with slightly-larger n.
%timeit -n1 naive_ways(35)

3.76 s ± 480 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [18]:
# Work for question A1
import numpy as np
a1_list = [2.8, 3.89, 5.77, 9.75, 16.2, 28.7]

itr = 1
slower_list = []
while itr < len(a1_list):
    slower_list.append(a1_list[itr] / a1_list[itr - 1])
    itr += 1

answer_a1 = np.asarray(slower_list).mean()
print("On average, ways(n) is {} times slower than ways(n−1) in the #TIMING INFO FOR PART A cell above".format(round(answer_a1, 2)))

On average, ways(n) is 1.6 times slower than ways(n−1) in the #TIMING INFO FOR PART A cell above


In [19]:
# Work for question A2
from collections import Counter
from collections import OrderedDict

class SortedDisplayDict(dict):
   def __str__(self):
       return "{" + ", ".join("%r: %r" % (key, self[key]) for key in sorted(self)) + "}"

def naive_ways_show_recusion(n, a):
    """Return the ways up a staircase of length n. Uses a naive algorithm."""
    if n == 1:
#         print(1)
        return 1
    if n == 2:
#         print(2)
        return 2
    if len(a) == 0:
        a.append(n)
    a.append(n-1)
    a.append(n-2)
#     print("call function", n-1, "call function", n-2)
    return naive_ways_show_recusion(n - 1, a) + naive_ways_show_recusion(n - 2, a)

a_3 = []
naive_ways_show_recusion(3, a_3)
print("n=3", SortedDisplayDict(Counter(a_3)))

a_4 = []
naive_ways_show_recusion(4, a_4)
print("n=4", SortedDisplayDict(Counter(a_4)))

a_5 = []
naive_ways_show_recusion(5, a_5)
print("n=5", SortedDisplayDict(Counter(a_5)))

a_6 = []
naive_ways_show_recusion(6, a_6)
print("n=6", SortedDisplayDict(Counter(a_6)))

a_7 = []
naive_ways_show_recusion(7, a_7)
print("n=7", SortedDisplayDict(Counter(a_7)))

a_8 = []
naive_ways_show_recusion(8, a_8)
print("n=8", SortedDisplayDict(Counter(a_8)))

a_9 = []
naive_ways_show_recusion(9, a_9)
print("n=9", SortedDisplayDict(Counter(a_9)))

a_25 = []
naive_ways_show_recusion(25, a_25)
print("n=25", SortedDisplayDict(Counter(a_25)))

n=3 {1: 1, 2: 1, 3: 1}
n=4 {1: 1, 2: 2, 3: 1, 4: 1}
n=5 {1: 2, 2: 3, 3: 2, 4: 1, 5: 1}
n=6 {1: 3, 2: 5, 3: 3, 4: 2, 5: 1, 6: 1}
n=7 {1: 5, 2: 8, 3: 5, 4: 3, 5: 2, 6: 1, 7: 1}
n=8 {1: 8, 2: 13, 3: 8, 4: 5, 5: 3, 6: 2, 7: 1, 8: 1}
n=9 {1: 13, 2: 21, 3: 13, 4: 8, 5: 5, 6: 3, 7: 2, 8: 1, 9: 1}
n=25 {1: 28657, 2: 46368, 3: 28657, 4: 17711, 5: 10946, 6: 6765, 7: 4181, 8: 2584, 9: 1597, 10: 987, 11: 610, 12: 377, 13: 233, 14: 144, 15: 89, 16: 55, 17: 34, 18: 21, 19: 13, 20: 8, 21: 5, 22: 3, 23: 2, 24: 1, 25: 1}


If we try to handle a staircase with 100 steps, it'll take a _very_ long time with our naive implementation.  (If you decide to try it, you'll want to interrupt your kernel instead of wait for the result!)

## A. Short Answer Questions

Give *brief* answers to the following in the cell below.

1. Based on the timing numbers from the `# TIMING INFO FOR PART A` cell above, approximately how much slower is $ways(n)$ than $ways(n-1)$? (i.e. what is $\frac{time(ways(n))}{time(ways(n-1))}$, roughly?)

2.  Why is this so slow?  What calculations do we compute repeatedly?  Hint: consider the following diagram.
![DP diagram](dp.png)

3. **Food for thought (not graded):** Assuming that $time(ways(n)) = time(ways(n-1)) + time(ways(n-2))$, what is $\lim_{n\to\infty} \frac{time(ways(n))}{time(ways(n-1))}$?

 A. Answers
1.  About 1.6 times slower
2.  If you are calling any whole number over n>3, you will be duplicating calls to recursive steps. If you look at the # Work for question A2 tab above, I run a bunch of naive_ways(n) functions with n: 3, 4, 5, 6, 7, 8, 9, and 25. What I do in those functions is append all of the recursive (n-1) and (n-2) that the naive_ways(n) calls within itself and aggregate the count on all the calls which are shown below in a dictionary (keys are step values, values are count of recursive call) and n= is the n value in the naive_ways(n) function:

n=3 {1: 1, 2: 1, 3: 1}

n=4 {1: 1, 2: 2, 3: 1, 4: 1}

n=5 {1: 2, 2: 3, 3: 2, 4: 1, 5: 1}

n=6 {1: 3, 2: 5, 3: 3, 4: 2, 5: 1, 6: 1}

n=7 {1: 5, 2: 8, 3: 5, 4: 3, 5: 2, 6: 1, 7: 1}

n=8 {1: 8, 2: 13, 3: 8, 4: 5, 5: 3, 6: 2, 7: 1, 8: 1}

n=9 {1: 13, 2: 21, 3: 13, 4: 8, 5: 5, 6: 3, 7: 2, 8: 1, 9: 1}

n=25 {1: 28657, 2: 46368, 3: 28657, 4: 17711, 5: 10946, 6: 6765, 7: 4181, 8: 2584, 9: 1597, 10: 987, 11: 610, 12: 377, 13: 233, 14: 144, 15: 89, 16: 55, 17: 34, 18: 21, 19: 13, 20: 8, 21: 5, 22: 3, 23: 2, 24: 1, 25: 1}

As you can see (for n>3), n and n-1 are called only once, but then after that n-2 and smaller (n-3, n-4, n-5....n-n+1) is recursivly called more than once and (n-3 and below) will be called more and more the larger the n input is for the naive_ways(n) function. This is because you are looping through all tree splits and redoing work that is already done. Lets take naive_ways(6) example. First this is the tree with looping through each branch and solving it (inefficient way):

![INEFFICIENT diagram](ineffcient_tree_solve.png)

And now lets look at what you could do if you used cache information to solve the tree:

![EFFICIENT diagram](effcient_tree_solve.png)

Efficiently all you need to do is solve for the right tree branch and save the values at each recursive n level.

Right branch side starting at bottom going up:
solve for:

** note n(2) = 2 and n(1) = 1

n(3) = n(2) + n(1) --> SOLVED

*plug into higher level

n(4) = n(3) + n(2) --> SOLVED

So now we have answers for n(1), n(2), n(3), n(4)

Go to the left branch and start at the bottom:

n(4) and n(3) --> already solved and in cache

n(5) = n(4) + n(3) --> solved

So now we have answers for n(1), n(2), n(3), n(4), and n(5).

Plug into the top level:

n(6) = n(5) + n(4) --> solved

We only calculated had to take the left branh down as far as the right branch top layer because all the values are already calculated and saved, so no need to loop through every branch of the tree to solve the problem.


3.  lim (n->inf) (n / n-1)
    
    **note: n = (n - 1) + (n - 2)
    
    = ((n - 1) + (n - 2)) / (n - 1) 
    
    = 1 + (n-2) / (n-1)
    
    n-1 / n-2 ~ 1.6  <-- solved in A1 above
    
    = 1 + (~1.6)^(-1)
    
    ~= 1.625
    



## Apply Dynamic Programming

Let's see if we can compute the results more cleverly, by keeping a cache (or table) of intermediate results that we can re-use.

First answer the following questions, then use what you gleaned to examine the code.

### B. Short Answer Questions
1. What are the values of A, B, and C in this table?
<html><table>
<tr><td>n</td><td>ways(n)</td></tr>
<tr><td>1</td><td>1</td></tr>
<tr><td>2</td><td>2</td></tr>
<tr><td>3</td><td>3</td></tr>
<tr><td>4</td><td>5</td></tr>
<tr><td>5</td><td>8</td></tr>
<tr><td>6</td><td>13</td></tr>
<tr><td>7</td><td>$A$</td></tr>
<tr><td>8</td><td>$B$</td></tr>
<tr><td>9</td><td>$C$</td></tr>
</table>
</html>
<p>
2. To compute these values, did you look at $n=4$ or earlier?
<p>
3. What is the minimum number of values you need to keep as you fill the table from top to bottom, while maintaining the DP property of not recomputing any values?  
<p>

### B. Answers

- 1A = 13 + 8 = 21
- 1B = 21 + 13 = 34
- 1C = 34 + 21 = 55
- 2 = No, no need to look at 4 or earlier to compute n = {7, 8, 9}. 

I solved:

(n=7) = (n=6) + (n=5)

(n=8) = (n=7) + (n=6)

(n=9) = (n=8) + (n=7)
- 3 = 2 values is the minimum needed. With the recursion f(n) = f(n-1) + f(n-2) and if you keep building up the table:

f(3) = f(2) + f(1)  --> keep f(3) and f(2) to calcultate f(4)

f(4) = f(3) + f(2)  --> keep f(4) and f(3) to calcultate f(5)

...

f(8) = f(7) + f(6)  --> finally keep f(8) and f(7) to calculate the last f(9)

f(9) = f(8) + f(7)

In [6]:
def dynamic_programming_ways(n):
    """Return the ways up a staircase of length n."""
    #  Initialize the cache to the answer for staircases of length one (1 way) and two (2 ways).
    cache = [1, 2]
    
    # Extend the cache, reusing the results we already have in the cache.
    # (For the mathematically inclined, you may see some
    #  parallels between DP and proof by induction.)
    for i in range(n - 2):
        cache.append(cache[i] + cache[i + 1])
    # We've extended the cache far enough that we can look up staircase of length n.
    # (Note that all the indices are one lower than the staircase, right from the
    #  time of initialization, hence the -1.)
    return cache[n - 1]

assert dynamic_programming_ways(1) == 1
assert dynamic_programming_ways(2) == 2
assert dynamic_programming_ways(100) == 573147844013817084101

%time dynamic_programming_ways(100)

Wall time: 0 ns


573147844013817084101

We can tweak things a little bit to reduce the memory footprint.  As seen in the short answer questions and in the code, you only need `cache[i-2]` and `cache[i-1]` to compute `cache[i]`.  Keeping earlier results from the cache are a bit of a waste.

In [7]:
def dynamic_programming_ways_efficient(n):
    """Return the ways up a staircase of length n."""
    #  Initialize the cache to the answer for staircases of length one (1 way) and two (2 ways).
    #  (We need to do some special casing here for n=1.  We could avoid this by initializing
    #   cache_i_1 == 1 as well and changing the for loop to be range(n-1).  Think about why...)
    if n == 1:
        return 1
    
    cache_i_2 = 1
    cache_i_1 = 2
    
    # Extend the cache, reusing the results we already have in the cache.
    # (For the mathematically inclined, you'll be excused if you see some
    #  parallels between DP and proof by induction.)
    for i in range(n - 2):
        new_value = cache_i_1 + cache_i_2
        cache_i_2 = cache_i_1
        cache_i_1 = new_value
        
    # We've extended the cache far enough that we can look up staircase of length n.
    # (Note that all the indices are one lower than the staircase, right from the
    #  time of initialization, hence the -1.)
    return cache_i_1

assert dynamic_programming_ways_efficient(1) == 1
assert dynamic_programming_ways_efficient(2) == 2
assert dynamic_programming_ways_efficient(100) == 573147844013817084101

%time dynamic_programming_ways_efficient(100)

Wall time: 0 ns


573147844013817084101

## Pipe cutting

Another problem suceptible to optimization with DP is pipe cutting.  Here, we are given a metal pipe of length $n$ and a price list, the price at which you can sell various parts of the pipe.  For the purposes of this exercise, let's write the code as though different parts of the pipe could yield different amounts of money, even if they're the same length.  Our objective is to cut the pipe into pieces such that they fetch the most revenue.

Similar to the stairs, we're going to slowly build a table up mapping the length of bars to the profit to be had.  If I have a bar of length 1, what's the most profitable way of cutting it up?  What about length 2?  Length 3? etc.  By keeping these intermediate results around, when we try cutting a length 10 pipe in one place, we can efficiently determine the profit made by optimally cutting the resulting left and right subpieces up.

Inspect the code below.

### Why do we care?

It turns out the pipe cutting problem is (very) closely related to DP algorithms for segmentation in NLP, as we'll see in more detail below. In Assignment 5, you'll implement the CKY algorithm for parsing, which is a sort of two-sided version of pipe-cutting that allows us to produce tree structures.

In [8]:
price_list = [0.0, -3.4, -5.7, 17.6, 2.2, 86.3]
def pipe_profit(left, right):
    """$ profit for the part of the bar in interval [left, right).
   
       Negative prices are a result of being unable to sell them
       and having to pay for disposal.  (Actually, as we'll learn
       later in the class, language models often produce -'ve scores,
       so we'll want to make sure our pipe splitting algorithm is robust
       to such scores when we use it in combination with a language model
       below.  Don't worry, you don't need to know anything about
       language models for this assignment.)
    """
    n = right - left 
    if n < 0 or n >= len(price_list):
        return 0.0
    return price_list[n]

# DP cache for the function below.
def best_cuts(n, score):
    """Determine the optimal revenue possible by (optionally) cutting a bar
       of length n.

    Args:
      n: the length of the pipe
      score: a function that accepts "left" and "right and gives you the score
             (revenue) received for the segment of pipe extending the interval
             from [left, right)
    
    Returns:
      The best profit to be had for a pipe of length n.
    """
        
    profit_on_left_most = []
    for k in range(n + 1):
        
        print("****************************************** k=", k, "******************************************")
        # What if we don't split the k-length bar at all?
        profit_on_left_most.append(score(0, k))

        # Maybe split it up.  Pick a split location at "j", optimally split the left
        # hand side but keep the right hand side whole.
        # (Answer below why this doesn't skip any options.)
        cut_list = []
        itr = 0
        for j in range(k):
            potential_profit = profit_on_left_most[j] + score(j, k)
            print(j,",", k, ", profit_on_left_most[j]=", profit_on_left_most[j], ", score(j, k)=", score(j, k), ", new total profit", potential_profit)
#             print(potential_profit, ",", profit_on_left_most[j], ",", score(j, k))
            if potential_profit > profit_on_left_most[k]:
                
                print("*****CUT MADE*****")
                profit_on_left_most[k] = potential_profit
                print(profit_on_left_most[k], "k-j=", k - j)
                cut_list.append(itr)
                itr = 0
            itr += 1
                
        print("\n")
        print("**** PIPE OPTIMIZED CUT LENGTH LIST ****")
        
        if len(cut_list) == 0:
            print("Optimal Pipe Length = ", k)
        else:
            if np.asarray(cut_list).sum() == k:
                print("Optimal Pipe Lengths were cut into:", "\n")
                for i in cut_list:
                    print(i, "\n")
            else:
                temp_num = k - np.asarray(cut_list).sum()
                cut_list.append(temp_num)
                for i in cut_list:
                    print(i, "\n")    
        print("****************************************** k=", k, "******************************************")
        print("\n")
    
#     print(profit_on_left_most)
    return profit_on_left_most[n]
   
profit_on_9 = round(best_cuts(9, pipe_profit), 1)
assert 100.5 == profit_on_9
print(profit_on_9)

# test = round(best_cuts(3, pipe_profit), 1)
# print(i, test)

****************************************** k= 0 ******************************************


**** PIPE OPTIMIZED CUT LENGTH LIST ****
Optimal Pipe Length =  0
****************************************** k= 0 ******************************************


****************************************** k= 1 ******************************************
0 , 1 , profit_on_left_most[j]= 0.0 , score(j, k)= -3.4 , new total profit -3.4


**** PIPE OPTIMIZED CUT LENGTH LIST ****
Optimal Pipe Length =  1
****************************************** k= 1 ******************************************


****************************************** k= 2 ******************************************
0 , 2 , profit_on_left_most[j]= 0.0 , score(j, k)= -5.7 , new total profit -5.7
1 , 2 , profit_on_left_most[j]= -3.4 , score(j, k)= -3.4 , new total profit -6.8


**** PIPE OPTIMIZED CUT LENGTH LIST ****
Optimal Pipe Length =  2
****************************************** k= 2 ******************************************


***

## C. Short Answer Questions

1. In the inner loop, why do we only need to try each location of a single cut (at `j`)?  Why don't we need to try cutting the bar into three or more pieces?

2. Answer the question posed in the code:  Why do we not need to optimally split both sides of `j`?  In other words, why can we get away with only taking the optimal split (`profit_on_left_most`) on the left and just `score()` the right?

3. Where are the cuts you should make for n=9?  Be sure to explain your work.  _(Hint:  Add a "print" to the code to see where you end up updating the profit made.  Leave this code in when you submit.)_

## C. Answers

1.  Since we started tracking information from initial length 1, 2, 3, ...., n we have all the information to solve the problem with only doing one cut. The reason is because we have figure out all the answers of the smaller lengths and can recursively solve the solution with one cut.

    For example lets say we have a hypothetical pipe length of 99. So the way we loop through the answers we have solved the ideal cut for length 1, 2, 3, .... all the way up to 98. So once we figure out where the first cut on the length 99 pipe is, we have the solution solved. 
    
    Hypothetical solution. First cut is made at 5. Leading to a pipe of length 5 and a pipe of length 94. We can use recursive information to solve the full solution. Now we plug in the answer we obtained from a length 5 into that and the solution we came up with for 94 into the length 5 one.
            
            - pipe 5 -> length 5 (i.e no cut)
            
            - pipe 94 -> cut into length 5 and 89
            
            ** now solve for 5, 89
            
            - pipe 89 -> cut into length 5 and 84
            
            ** solve ....
            
            ....
            
            - until the final solution with the recursion algorithm we chose to follow

2.  We are only looking at total length of the cut pipe segments for profit maximization. There is no coupling of price difference of actual location of the cut. Since it is just total length of each pipe segment, if we were to do both sides, we would just be looping through the data over and over for no added information. Starting on one side and adding length will take into account recursive information of smaller segments and build the information up to larger and larger lengths of pipes.

    Ex. ---------
    Looking at length 4 information will require one loop:

    (1) : ----|-----

    However if you did it on both sides you will get information for:

    start left to right

    (l1) : |----|-----

    (l2) : -|----|----

    (l3) : --|----|---

    (l4) : ---|-----|-

    (l5) : ----|-----|

    finish left to right

    start right to left

    (r1) : ----|-----|

    (r2) : ---|-----|-

    (r3) : --|----|---

    (r4) : -|----|----

    (r5) : |----|-----

    end left to right

    10 loops for length-4 segment information that gives you the same profit for each case for the length-4 bar. In addition, you are forced into certain pipe configurations if certain patterns are used. For example if you used (l3) or (r3) configuration then you are forced to have length-2 pipe on one side and length-3 on the other where:

    length-2 pipe can be:
    
          1 length-2 pipe 
  
          2 length-1 pipes

    length-3 pipe can be: 

          1 length-3 pipe 

          3 length-1 pipes

          1 length-1 pipe and 1 length-2 pipe

    With that logic you have constrained yourself too much for the solution. In contrast if you use (1) logic then you are left with a segment length-5 which has the capability to be split up into many more configurations segment lengths in order to optimize profit based on length of pipes only.

    length-5 pipe can be:

 1 length-5 pipe
 1 length-4 pipe and 1 length 1 pipe
 ....
 5 length-1 pipes
 
 One loop for cut on left and score on right vs looping through every single branch of combinations to come up with the same answer (most optimized solution).
     
3.  cuts were made to make the n=9 pipe into a pipe with three segments with sizes 1, 3, and 5:

        Actual cut into length 3 and length 6
        
            length 3 optimimal length is 3

            length 6 optimal is length 5 and a length 1

                    length 5 optimal length is 5

                    length 1 optimal length is 1

## Bookkeeping

Great!  We know we can make over a hundred dollars cutting up our length 9 bar!

Unfortunately, while we computed the revenue available by cutting the bar, we didn't actually track the cuts we need to make in order to earn it!

## Coding Exercise
**Finish modifying the code (in the cell below)** to keep track of the cuts you made in order to achieve the optimal revenue.

i.e. `potential_profit[4]` might be `(39.5, [1, 2])`.

In [99]:
price_list = [0.0, -3.4, -5.7, 17.6, 2.2, 86.3]
def pipe_profit(left, right):
    """$ profit for the part of the bar in interval [left, right).
   
       Negative prices are a result of being unable to sell them
       and having to pay for disposal.  (Actually, as we'll learn
       later in the class, language models often produce -'ve scores,
       so we'll want to make sure our pipe splitting algorithm is robust
       to such scores when we use it in combination with a language model
       below.  Don't worry, you don't need to know anything about
       language models for this assignment.)
    """
    n = right - left 
    if n < 0 or n >= len(price_list):
        return 0.0
    return price_list[n]

# DP cache for the function below.
def best_cuts(n, score):
    """Determine the optimal revenue possible by (optionally) cutting a bar
       of length n.

    Args:
      n: the length of the pipe
      score: a function that accepts "left" and "right and gives you the score
             (revenue) received for the segment of pipe extending the interval
             from [left, right)
    
    Returns:
      The best profit to be had for a pipe of length n.
    """
    
    
    profit_on_left_most = []
    segment_lengths_per_pipe = {}
    for k in range(n + 1):
#         print("************** ", k, " **************")
        if k > 0:  
            segment_lengths_per_pipe[k] = {}
            segment_lengths_per_pipe[k]['pipe_segments'] = []
            segment_lengths_per_pipe[k]['profit'] = {}
        profit_on_left_most.append(score(0, k))
        # Maybe split it up.  Pick a split location at "j", optimally split the left
        # hand side but keep the right hand side whole.
        # (Answer below why this doesn't skip any options.)
        cut_list = []
        itr = 0
        for j in range(k):
            potential_profit = profit_on_left_most[j] + score(j, k)
#             print(j,",", k)
            if potential_profit > profit_on_left_most[k]:
#                 print("*****CUT MADE*****")
                if segment_lengths_per_pipe[itr] and len(segment_lengths_per_pipe[k]['pipe_segments']) <= 2:
                    segment_lengths_per_pipe[k]['pipe_segments'].append(itr)
                    segment_lengths_per_pipe[k]['pipe_segments'] = segment_lengths_per_pipe[k]['pipe_segments'] + segment_lengths_per_pipe[k - itr]['pipe_segments']
                profit_on_left_most[k] = potential_profit   
            itr += 1
            if itr >= k and len(segment_lengths_per_pipe[k]['pipe_segments']) == 0:
                segment_lengths_per_pipe[k]['pipe_segments'].append(k)
#         print("************** ", k, " **************")
        
        if k > 0:  
            segment_lengths_per_pipe[k]['profit'] = profit_on_left_most[k]
        
    
    return segment_lengths_per_pipe

        
segment_lengths_per_pipe = best_cuts(50, pipe_profit)

for i in segment_lengths_per_pipe.keys():
    if i <= 9:
        print("For initial pipe length: {}, cut pipe segments length (optimized): {}, for a total profit of: {}".format(i, segment_lengths_per_pipe[i]['pipe_segments'], segment_lengths_per_pipe[i]['profit']))

For initial pipe length: 1, cut pipe segments length (optimized): [1], for a total profit of: -3.4
For initial pipe length: 2, cut pipe segments length (optimized): [2], for a total profit of: -5.7
For initial pipe length: 3, cut pipe segments length (optimized): [3], for a total profit of: 17.6
For initial pipe length: 4, cut pipe segments length (optimized): [1, 3], for a total profit of: 14.200000000000001
For initial pipe length: 5, cut pipe segments length (optimized): [5], for a total profit of: 86.3
For initial pipe length: 6, cut pipe segments length (optimized): [1, 5], for a total profit of: 82.89999999999999
For initial pipe length: 7, cut pipe segments length (optimized): [2, 5], for a total profit of: 80.6
For initial pipe length: 8, cut pipe segments length (optimized): [3, 5], for a total profit of: 103.9
For initial pipe length: 9, cut pipe segments length (optimized): [3, 1, 5], for a total profit of: 100.5


We'll see bookkeeping like this throughout the rest of the course.  For example, we'll want to know the optimal way to tag words in a sentence with their parts of speech.  The optimization will be over some likelihood of a particular assignment (rather than revenue).  This optimization is only useful to us however if we have a way to know what sequence of part of speech tags gave us that score, so we'll have to do this same kind of bookkeeping.

## Pipe cutting is segmentation
The next cell implements a very light wrapper around your best_cuts_with_trace function (it just takes the cut indexes your function returns and turns them into text to pretty-print).  It also implements a simple unigram language model (much simpler than what you will build in later assignments!). 

Don't worry about the details here, this is just a fun coda to the assignment!  Figuring out these details will be the work of the rest of the course.

Feel free to experiment with some sentences.  See if you can find at least one that breaks it, yielding a sub-optimal segmentation. If the dynamic programming algorithm is exact (finds the highest scoring split), how can it produce a bad solution?

In [100]:
import numpy as np

##
# Compute unigram counts from a simple corpus.
unigram_counts = {}
total_counts = 0
for line in open('english_uni_simplified_sorted_top', encoding="utf8").readlines():
    word_and_count = line.split()
    word = word_and_count[0].strip('"')
    count = int(word_and_count[1])
    unigram_counts[word] = count
    total_counts += count

def unigram_scoring_function(text, left, right):
    word = text[left:right]
    if word in unigram_counts:
        # Log probabilities, so we can add scores instead of multiplying
        return np.log(unigram_counts[word]) - np.log(total_counts)
    else:  
        # "Smoothing", encouraging in-vocabulary, or at least short OOV words.
        # We give a lower score to longer out-of-vocabulary spans.
        return -100 * (right - left)
        
##
# Use the pipe-cutting algorithm to segment text.
def segment(text, segment_lengths_per_pipe):
    # We create a scoring lambda that accepts two parameters, "left" and "right", as required by the
    # code you implemented above.  However, we also need access to the "text" in order
    # to score the unigram.  A lambda captures the local variable "text" for this purpose.
    score_func = lambda left, right: unigram_scoring_function(text, left, right)
    
    # Call your function to slice the string.
#     score, cuts = best_cuts_with_trace(len(text), score_func)

    cuts = segment_lengths_per_pipe[len(text)]['pipe_segments']
    score = segment_lengths_per_pipe[len(text)] ['profit']
    
    # Imply a "cut" at the start and end of the text so that the list comprehension below is convenient.
    itr = 0
    for i in cuts:
        if itr != 0:
            cuts[itr] = cuts[itr] + cuts[itr - 1]
        itr += 1
    if cuts[0] != 0:
        cuts.insert(0, 0)            
    
    print(cuts)
    str_out = list(filter(None, [text[cuts[i] : cuts[i + 1]] for i in range(len(cuts) - 1)]))
    
    # Convert the list of cuts into a list of words.
    return score, str_out

In [102]:
segment('helloworldhowareyou', segment_lengths_per_pipe)

[0, 3, 9, 18, 30, 44, 63]


(273.1, ['hel', 'loworl', 'dhowareyo', 'u'])

In [103]:
segment('downbythebay', segment_lengths_per_pipe)

[0, 3, 6, 7, 12]


(166.9, ['dow', 'nby', 't', 'hebay'])

In [104]:
segment('wikipediaisareallystrongresourceontheinternet', segment_lengths_per_pipe)

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 40, 45]


(776.6999999999998,
 ['wik',
  'ipe',
  'dia',
  'isa',
  'rea',
  'lly',
  'str',
  'ong',
  'res',
  'our',
  'ceo',
  'nth',
  'ein',
  't',
  'ernet'])

## Congratulations!

You're done with Dynamic Programming.

There is a completely optional section that shouldn't take very long, if you're keen to learn about edit distance.  We won't delve into much detail about this anywhere else in the class.

## (Optional) String edit distance

Another classic DP problem in the NLP space - but not one we otherwise will talk about in the course is the idea of "[edit distance](https://en.wikipedia.org/wiki/Levenshtein_distance)".  It's a way of measuring how many "edits" to one string you need to make in order to turn it into another.

We've provided two implementations below for you to play with.

1.  **levenshtein_cache:** The "cache everything in a dict" approach is first.  The keys are coordinates into a table that is len(str1) x len(str2) in size.

2.  **levenshtein_explicit:** Similar to the version of ways(n) that only keeps the previous two values at hand, the explicit ordering approach only keeps the immediately previous row of the table while building the next. Setting the verbose flag to this version prints each row of the table out as it computes it.

## D. (Optional) Short Answer Questions:

Give brief answers to the following in the cell below.

1. Let `n = len(str1)` and `m = len(str2)`. In terms of `n` and `m`, what is the size of the DP table (cache) for computing Levenshtein distance? _Hint: how many valid keys are there? Do we use all of them?_
<p>
2. Based on your answer to 1., what is the running time (in Big-O notation) of the edit distance algorithm? _Hint: it takes $O(1)$ work at each step, assuming we have the needed cache entries._
<p>

3. Consider transpositions (as mentioned in section 5.6 of the async), such as `xy` -> `yx`. How can we compose a transposition from insertions, deletions, and substitutions? What is the edit distance between `wxyz` and `wyxz`?
<p>

4. Suppose we wanted to handle transpositions directly, rather than allowing our algorithm to compose them from other operations. (This might be useful if we want to score them differently.) If we have for the other operations:
```python
_ed(i - 1, j) + 1  # insertion
_ed(i, j - 1) + 1  # deletion
_ed(i - 1, j - 1) + substitution  # substitution, free if letters match  
```
what line would we add (calling `_ed`) to handle a transposition? (You may want to define a variable `transposition_match` to check that a transposition makes sense at the current position.) Based on your answer to 1. and 2., does this change the Big-O runtime of the algorithm?

## D. Answers (optional)

1. _Your answer here!_
2. _Your answer here!_
3. _Your answer here!_
4. _Your answer here!_

In [7]:
def levenshtein_cache(str1, str2):
    cache = dict()
    def _ed(i, j):
        """Recursive helper, using cache."""
        if (i,j) in cache: 
            return cache[(i,j)]
        
        # Base cases
        if i == 0:
            result = j
        elif j == 0:
            result = i
            
        # Main recursion
        else:
            # 1 if letters differ (substitution is free if the letters are the same)
            substitution = 0 if str1[i - 1] == str2[j - 1] else 1
            result = min([
                    _ed(i - 1, j) + 1,  # insertion
                    _ed(i, j - 1) + 1,  # deletion
                    _ed(i - 1, j - 1) + substitution  # substitution, free if letters match  
            ])
        cache[(i,j)] = result
        return result
    
    return _ed(len(str1), len(str2))

In [8]:
def levenshtein_explicit(str1, str2, verbose=False):
    prev_num_edits = range(len(str1) + 1)
    for j in range(1, len(str2) + 1):
        num_edits = [prev_num_edits[0] + 1]
        for i in range(1, len(str1) + 1):
            # 1 if letters differ (substitution is free if the letters are the same)
            substitution = 0 if str1[i - 1] == str2[j - 1] else 1
            result = min([num_edits[i - 1] + 1,
                          prev_num_edits[i] + 1,
                          prev_num_edits[i - 1] + substitution
            ])
            num_edits.append(result)
        if verbose:
            print(prev_num_edits)
        prev_num_edits = num_edits
    if verbose:
        print(prev_num_edits)
    return prev_num_edits[len(str1)]

In [9]:
# Substitution.
levenshtein_explicit('abc', 'dbc', verbose=True)

In [10]:
# Deletion.
levenshtein_explicit('abc', 'ac')

In [11]:
# Insertion.
levenshtein_explicit('ac', 'abc')

In [12]:
# All of the above.
levenshtein_cache('kitten', 'sitting')

In [13]:
# Fun!
levenshtein_cache('w266 class', 'with 6 classic tricks')